File: mobile/lib/features/panic_alert/domain/entities/alert_entity.dart
import 'package:equatable/equatable.dart';

/// Entidad de Alerta de Pánico — capa de Domain (Dart puro).
///
/// Compliance (Informe #1):
/// - [photoUrl] es una URL firmada de S3 con expiración de 90 días.
/// - [latitude] y [longitude] son datos sensibles — nunca
///   mostrar la ubicación exacta a vecinos, solo zona aproximada.
/// - Esta entidad NUNCA debe tener referencias a Flutter widgets.
class AlertEntity extends Equatable {
  const AlertEntity({
    required this.id,
    required this.userId,
    required this.latitude,
    required this.longitude,
    required this.status,
    required this.triggeredAt,
    this.photoUrl,
    this.resolvedAt,
    this.respondersCount = 0,
  });

  final String id;
  final String userId;

  /// Ubicación exacta — dato sensible (Compliance).
  /// Solo compartir zona aproximada con vecinos.
  final double latitude;
  final double longitude;

  final AlertStatus status;
  final DateTime triggeredAt;

  /// URL firmada S3 (expira en 90 días) — dato crítico (Compliance).
  final String? photoUrl;

  final DateTime? resolvedAt;
  final int respondersCount;

  bool get isActive => status == AlertStatus.active;
  bool get canBeCancelled =>
      isActive &&
      DateTime.now().difference(triggeredAt).inSeconds <= 30;

  @override
  List<Object?> get props => [
        id,
        userId,
        latitude,
        longitude,
        status,
        triggeredAt,
        photoUrl,
        resolvedAt,
        respondersCount,
      ];
}

enum AlertStatus {
  active,
  resolved,
  falseAlarm,
  cancelled;

  bool get isActive => this == AlertStatus.active;
}

File: backend/src/modules/alerts/entities/alert.entity.ts
import { Entity, Column, PrimaryGeneratedColumn, CreateDateColumn } from 'typeorm';

@Entity('panic_alerts')
export class AlertEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ name: 'user_id' })
  userId: string;

  @Column('double precision')
  latitude: number;

  @Column('double precision')
  longitude: number;

  @Column({ name: 'photo_url', nullable: true })
  photoUrl: string | null;

  @Column({ default: 'active' })
  status: 'active' | 'resolved' | 'false_alarm' | 'cancelled';

  @CreateDateColumn({ name: 'triggered_at' })
  triggeredAt: Date;

  @Column({ name: 'resolved_at', nullable: true })
  resolvedAt: Date | null;
}
